File: 백준_3 Gold/21761_초직사각형.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

vector<long long> hyperCube(4);
vector<vector<long long>> cards(4);
vector<int> index(4, 0);

int getLargestDim();

int main() {
    int N, K;
    cin >> N >> K;
    for (int i = 0; i < 4; i++)
        cin >> hyperCube[i];
    
    char T;
    int U;
    for (int i = 0; i < N; i++) {
        cin >> T >> U;
        cards[T - 'A'].push_back(U);
    }

    for(int i = 0; i < 4; i++)
        sort(cards[i].rbegin(), cards[i].rend());
    
    for(int i = 0; i < K; i++){
        int curIncDim = getLargestDim();
        cout << (char)('A' + curIncDim) << ' ' << cards[curIncDim][index[curIncDim]] << '\n';
        hyperCube[curIncDim] += cards[curIncDim][index[curIncDim]];
        index[curIncDim] += 1;
    }

}

int getLargestDim(){

    long long maxIncreaseVolume = 0;
    int maxIncVolInd;
    
    for(int i = 0; i < 4; i++){
        if(cards[i].size() == index[i]) continue;

        long long curVolIncrease = cards[i][index[i]];
        for(int j = 0; j < 4; j++){
            if(j == i) continue;
            curVolIncrease *= hyperCube[j];
        }

        if(maxIncreaseVolume < curVolIncrease){
            maxIncreaseVolume = curVolIncrease;
            maxIncVolInd = i;
        }
    }
    
    return maxIncVolInd;
}

File: 백준_3 Gold/21761_초직사각형.py
cards = [[], [], [], []]
index = [0, 0, 0, 0]

def getLargestDim():

    maxIncreaseVolume = 0
    maxIncVolInd = -1
    
    for i in range(4):
        if len(cards[i]) == index[i]:
            continue

        curVolIncrease = cards[i][index[i]]
        for j in range(4):
            if j == i:
                continue;
            curVolIncrease *= hyperCube[j]

        if maxIncreaseVolume < curVolIncrease:
            maxIncreaseVolume = curVolIncrease
            maxIncVolInd = i
    
    return maxIncVolInd

N, K = map(int, input().split(' '))
hyperCube = list(map(int, input().split(' ')))

for i in range(N):
    T, U = input().split()
    cards[ord(T) - ord('A')].append(int(U));

for i in range(4):
    cards[i].sort(reverse = True)

for i in range(K):
    curIncDim = getLargestDim();
    print(chr(ord('A') + curIncDim), cards[curIncDim][index[curIncDim]])
    hyperCube[curIncDim] += cards[curIncDim][index[curIncDim]];
    index[curIncDim] += 1;
